presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java

```diff
@@ -126,6 +142,50 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
         }
     }
 
+    public HttpServletRequest mergeExtraHeaders(HttpServletRequest request, Principal principal)
+    {
+        List<ClientRequestFilter> clientRequestFilters = clientRequestFilterManager.getClientRequestFilters();
+
+        if (clientRequestFilters.isEmpty()) {
+            return request;
+        }
+
+        ImmutableMap.Builder<String, String> extraHeadersMapBuilder = ImmutableMap.builder();
+        Set<String> addedHeaders = new HashSet<>();
+
+        for (ClientRequestFilter requestFilter : clientRequestFilters) {
+            boolean headersPresent = requestFilter.getHeaderNames().stream()
+                    .allMatch(headerName -> request.getHeader(headerName) != null);
+
+            if (!headersPresent) {
+                ClientRequestFilterContext context = new ClientRequestFilterContext(principal);
+                Optional<Map<String, String>> extraHeaderValueMap = requestFilter.getExtraHeaders(context);
+
+                if (extraHeaderValueMap.isPresent()) {
+                    Map<String, String> extraHeaderValues = extraHeaderValueMap.get();
+                    for (Map.Entry<String, String> extraHeaderEntry : extraHeaderValues.entrySet()) {
+                        String headerKey = extraHeaderEntry.getKey();
+                        if (headersBlockList.contains(headerKey)) {
+                            throw new PrestoException(HEADER_MODIFICATION_ATTEMPT,
+                                    "Modification attempt detected: The header " + headerKey + " is not allowed to be modified. The following headers cannot be modified: " +
+                                            String.join(", ", headersBlockList));
+                        }
+                        if (addedHeaders.contains(headerKey)) {
+                            throw new PrestoException(HEADER_MODIFICATION_ATTEMPT, "Header conflict detected: " + headerKey + " already added by another filter.");
+                        }
+                        if (request.getHeader(headerKey) == null && requestFilter.getHeaderNames().contains(headerKey)) {
+                            extraHeadersMapBuilder.put(headerKey, extraHeaderEntry.getValue());
+                            addedHeaders.add(headerKey);
+                        }
+                    }
+                }
+            }
+        }
+
+        Map<String, String> extraHeadersMap = extraHeadersMapBuilder.build();
+        return new ModifiedHttpServletRequest(request, extraHeadersMapBuilder.build());
```

Comment:
```suggestion
        return new ModifiedHttpServletRequest(request, extraHeadersMapBuilder.build());
```

Remove duplicate build.